server/controllers/wip/questionsController.ts

```diff
@@ -92,6 +97,85 @@ export default class QuestionsController extends BaseController<FormWizard.Multi
     })
   }
 
+  async saveValues(
+    req: FormWizard.Request<FormWizard.MultiValues, string>,
+    res: express.Response,
+    next: express.NextFunction,
+  ): Promise<void> {
+    const { incidentReportingApi } = res.locals.apis
+
+    const submittedValues = req.form.values
+
+    const report = res.locals.report as ReportWithDetails
+    const reportConfig = res.locals.reportConfig as IncidentTypeConfiguration
+
+    const questionsResponses = []
+    for (const [fieldName, values] of Object.entries(submittedValues)) {
+      const questionConfig = reportConfig.questions[fieldName]
+      if (questionConfig === undefined) {
+        logger.error(
+          new Error(
+            `Report '${report.id}': Submitted Field '${fieldName}' not found in ${report.type}'s configuration.`,
+          ),
+        )
+
+        // eslint-disable-next-line no-continue
+        continue
+      }
+      const questionResponses: AddOrUpdateQuestionWithResponsesRequest = {
+        code: fieldName,
+        question: questionConfig.code,
+        additionalInformation: null,
+        responses: [],
+      }
+
+      const responseCodes = (questionConfig.multipleAnswers ? values : [values]) as string[]
+      for (const responseCode of responseCodes) {
+        if (responseCode === '') {
+          // eslint-disable-next-line no-continue
+          continue
+        }
+        const answerConfig = this.findAnswerConfigByCode(responseCode, questionConfig)
+        if (answerConfig === undefined) {
+          logger.error(
+            new Error(
+              `Report '${report.id}': Submitted Answer with code '${responseCode}' not found in ${report.type}'s question '${questionConfig.id}' configuration.`,
+            ),
+          )
+          // eslint-disable-next-line no-continue
+          continue
+        }
+
+        const response: AddOrUpdateQuestionResponseRequest = {
+          response: responseCode,
+          responseDate: null,
+          additionalInformation: null,
+        }
+
+        if (answerConfig.commentRequired) {
+          const commentFieldName = `${questionConfig.id}-${answerConfig.id}-comment`
+          if (submittedValues[commentFieldName] !== '') {
+            response.additionalInformation = submittedValues[commentFieldName] as string
+          }
+        }
+
+        if (answerConfig.dateRequired) {
+          const dateFieldName = `${questionConfig.id}-${answerConfig.id}-date`
+          if (submittedValues[dateFieldName] !== '') {
+            response.responseDate = parseDateInput(submittedValues[dateFieldName] as string)
+          }
+        }
+        questionResponses.responses.push(response)
+      }
+
+      questionsResponses.push(questionResponses)
+    }
+
+    await incidentReportingApi.addOrUpdateQuestionsWithResponses(report.id, questionsResponses)
+
+    super.saveValues(req, res, next)
```

Comment:
it would you just gotta wrap their next handler with a new one like:

```typescript
super.saveValues(req, res, error => {
  if (error) {
    next(error)
  } else {
    // your code
    next()
  }
})
```

Reply:
It's a bit ugly but I can give it a try :)

Reply:
yes, the days before async/await were ugly

Reply:
Indeed. Updated here: https://github.com/ministryofjustice/hmpps-incident-reporting/pull/281/commits/b740b685e7ebf7a8b4569585fd47e767cad7dad1